checks/net.go

```diff
@@ -30,43 +50,73 @@ func (c *ConnectionsCheck) Endpoint() string { return "/api/v1/collector" }
 func (c *ConnectionsCheck) RealTime() bool { return false }
 
 // Run runs the ConnectionsCheck to collect the live TCP connections on the
-// system. In most POSIX systems we will use the procfs net files to read out
-// this information. For each connection we'll return a `model.Connection` that
-// will be bundled up into a `CollectorConnections`.
+// system. Currently only linux systems are supported as eBPF is used to gather
+// this information. For each connection we'll return a `model.Connection`
+// that will be bundled up into a `CollectorConnections`.
 // See agent.proto for the schema of the message and models.
 func (c *ConnectionsCheck) Run(cfg *config.AgentConfig, groupID int32) ([]model.MessageBody, error) {
-	start := time.Now()
-	connections, err := net.ConnectionsMax("tcp", cfg.MaxProcFDs)
-	if err != nil && err.Error() == util.ErrNotImplemented.Error() {
+	if !c.supported || c.tracer == nil {
 		return nil, nil
-	} else if err != nil {
+	}
+
+	start := time.Now()
+
+	conns, err := c.tracer.GetActiveConnections()
+	if err != nil {
+		if err == tracer.ErrNotImplemented {
+			return nil, nil
+		}
 		return nil, err
 	}
 
-	log.Infof("collected connections in %s", time.Now().Sub(start))
+	log.Infof("collected connections in %s", time.Since(start))
 	return []model.MessageBody{&model.CollectorConnections{
 		HostName:    cfg.HostName,
-		Connections: formatConnections(connections),
+		Connections: formatConnections(conns),
 	}}, nil
 }
 
-func formatConnections(stats []net.ConnectionStat) []*model.Connection {
-	cxs := make([]*model.Connection, 0, len(stats))
-	for _, c := range stats {
+// TODO: Break up large connection messages into batches
+func formatConnections(conns []tracer.ConnectionStats) []*model.Connection {
+	cxs := make([]*model.Connection, 0, len(conns))
+	for _, c := range conns {
 		cxs = append(cxs, &model.Connection{
 			Pid:    int32(c.Pid),
-			Fd:     int32(c.Fd),
-			Family: int32(c.Family),
-			Type:   int32(c.Type),
+			Family: int32(formatFamily(c.Family)),
+			Type:   int32(formatType(c.Type)),
 			Laddr: &model.Addr{
-				Ip:   c.Laddr.IP,
-				Port: int32(c.Laddr.Port),
+				Ip:   c.Source,
+				Port: int32(c.SPort),
 			},
 			Raddr: &model.Addr{
-				Ip:   c.Raddr.IP,
-				Port: int32(c.Raddr.Port),
+				Ip:   c.Dest,
+				Port: int32(c.DPort),
 			},
+			BytesSent:     int64(c.SendBytes), // TODO: Consider sending rate instead of cumulative byte counts
+			BytesRecieved: int64(c.RecvBytes),
```

Comment:
I think moving to rates might be worth doing from the get-go as it makes the backend processing all the more easier. we can technically support rates with metrics (we do this for one metric now) but many other metrics from this and the infra agent are sent as rates.